File: lint-arwaky/src-python/taxonomy/lint_adapter_error.py
"""lint_adapter_error — Linting domain error types."""

from pydantic import BaseModel, ConfigDict, Field

from .error_code_vo import ErrorCode
from .error_value_vo import (
    ErrorMessage,
    ExitCode,
    FieldName,
    Constraint,
    ActualValue,
    Cause,
)
from .file_path_vo import FilePath
from .adapter_name_vo import AdapterName
from .lint_domain_vo import CommandArgs


class AdapterError(BaseModel):
    """External linter tool failed."""

    model_config = ConfigDict(frozen=True)

    adapter_name: AdapterName
    message: ErrorMessage
    error_code: ErrorCode | None = None
    command: CommandArgs = CommandArgs()
    stderr: ErrorMessage = Field(default_factory=lambda: ErrorMessage(value=""))
    exit_code: ExitCode | None = None

    def __str__(self):
        code = f" [{self.error_code}]" if self.error_code else ""
        return f"[{self.adapter_name}]{code} {self.message}"


class ScanError(BaseModel):
    """Scan operation failed on a path."""

    model_config = ConfigDict(frozen=True)

    path: FilePath
    message: ErrorMessage
    error_code: ErrorCode | None = None
    adapter_name: AdapterName | None = None
    cause: Cause | None = None

    def __str__(self):
        adapter = f" ({self.adapter_name})" if self.adapter_name else ""
        code = f" [{self.error_code}]" if self.error_code else ""
        return f"Scan failed{adapter}{code}: {self.path} — {self.message}"


class ValidationError(BaseModel):
    """Input validation failed."""

    model_config = ConfigDict(frozen=True)

    field_name: FieldName
    message: ErrorMessage
    constraint: Constraint = Field(default_factory=lambda: Constraint(value=""))
    value: ActualValue = Field(default_factory=lambda: ActualValue(value=""))

    def __str__(self):
        return f"Validation failed on '{self.field_name}': {self.message}"

File: lint-arwaky/src-rust/taxonomy/lint_adapter_error.rs
use serde::{Serialize, Deserialize};
use super::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AdapterError {
    pub adapter_name: AdapterName,
    pub message: ErrorMessage,
    #[serde(default)]
    pub error_code: Option<ErrorCode>,
    #[serde(default)]
    pub command: Option<String>,
    #[serde(default)]
    pub stderr: Option<ErrorMessage>,
    #[serde(default)]
    pub exit_code: Option<ExitCode>,
}

impl AdapterError {
    pub fn new(adapter_name: AdapterName, message: ErrorMessage) -> Self {
        Self { adapter_name, message, error_code: None, command: None, stderr: None, exit_code: None }
    }
}

impl std::fmt::Display for AdapterError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let code = self.error_code.as_ref().map(|c| format!(" [{}]", c)).unwrap_or_default();
        write!(f, "[{}]{} {}", self.adapter_name, code, self.message)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ScanError {
    pub path: FilePath,
    pub message: ErrorMessage,
    #[serde(default)]
    pub error_code: Option<ErrorCode>,
    #[serde(default)]
    pub adapter_name: Option<AdapterName>,
    #[serde(default)]
    pub cause: Option<Cause>,
}

impl ScanError {
    pub fn new(path: FilePath, message: ErrorMessage) -> Self {
        Self { path, message, error_code: None, adapter_name: None, cause: None }
    }
}

impl std::fmt::Display for ScanError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let adapter = self.adapter_name.as_ref().map(|a| format!(" ({})", a)).unwrap_or_default();
        let code = self.error_code.as_ref().map(|c| format!(" [{}]", c)).unwrap_or_default();
        write!(f, "Scan failed{}{}: {} — {}", adapter, code, self.path, self.message)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ValidationError {
    pub field_name: String,
    pub message: ErrorMessage,
    #[serde(default)]
    pub constraint: Option<String>,
    #[serde(default)]
    pub value: Option<String>,
}

impl ValidationError {
    pub fn new(field_name: String, message: ErrorMessage) -> Self {
        Self { field_name, message, constraint: None, value: None }
    }
}

impl std::fmt::Display for ValidationError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Validation failed on '{}': {}", self.field_name, self.message)
    }
}
